packages/tools/src/tools/segmentation/CircleROIStartEndThresholdTool.ts

```diff
@@ -0,0 +1,625 @@
+  constructor(
+    toolProps: PublicToolProps = {},
+    defaultToolProps: ToolProps = {
+      supportedInteractionTypes: ['Mouse', 'Touch'],
+      configuration: {
+        numSlicesToPropagate: 10,
+      },
+    }
+  ) {
+    super(toolProps, defaultToolProps);
+
+    this._throttledCalculateCachedStats = throttle(
+      this._calculateCachedStatsTool,
+      100,
+      { trailing: true }
+    );
+  }
+
+  /**
+   * Based on the current position of the mouse and the current imageId to create
+   * a CircleROI Annotation and stores it in the annotationManager
+   *
+   * @param evt -  EventTypes.NormalizedMouseEventType
+   * @returns The annotation object.
+   *
+   */
+  addNewAnnotation = (evt: EventTypes.InteractionEventType) => {
+    const eventDetail = evt.detail;
+    const { currentPoints, element } = eventDetail;
+    const worldPos = currentPoints.world;
+    const canvasPos = currentPoints.canvas;
+
+    const enabledElement = getEnabledElement(element);
+    const { viewport, renderingEngine } = enabledElement;
+
+    this.isDrawing = true;
+
+    const camera = viewport.getCamera();
+    const { viewPlaneNormal, viewUp } = camera;
+
+    let referencedImageId, imageVolume, volumeId;
+    if (viewport instanceof StackViewport) {
+      throw new Error('Stack Viewport Not implemented');
+    } else {
+      const targetId = this.getTargetId(viewport);
+      volumeId = targetId.split('volumeId:')[1];
+      imageVolume = cache.getVolume(volumeId);
+      referencedImageId = csUtils.getClosestImageId(
+        imageVolume,
+        worldPos,
+        viewPlaneNormal
+      );
+    }
+
+    if (!referencedImageId) {
+      throw new Error('This tool does not work on non-acquisition planes');
+    }
+
+    const spacingInNormal = csUtils.getSpacingInNormalDirection(
+      imageVolume,
+      viewPlaneNormal
+    );
+
+    const newStartIndex = this._getStartSliceIndex(
+      imageVolume,
+      worldPos,
+      spacingInNormal,
+      viewPlaneNormal
+    );
+
+    // We cannot newStartIndex add numSlicesToPropagate to startIndex because
+    // the order of imageIds can be from top to bottom or bottom to top and
+    // we want to make sure it is always propagated in the direction of the
+    // view and also to make sure we don't go out of bounds.
+    const endIndex = this._getEndSliceIndex(
+      imageVolume,
+      worldPos,
+      spacingInNormal,
+      viewPlaneNormal
+    );
+
+    console.log(endIndex);
+
+    const FrameOfReferenceUID = viewport.getFrameOfReferenceUID();
+
+    const annotation = {
+      highlighted: true,
+      invalidated: true,
+      metadata: {
+        toolName: this.getToolName(),
+        viewPlaneNormal: <Types.Point3>[...viewPlaneNormal],
+        viewUp: <Types.Point3>[...viewUp],
+        FrameOfReferenceUID,
+        referencedImageId,
+        volumeId,
+        spacingInNormal,
+        enabledElement,
+      },
+      data: {
+        label: '',
+        startSlice: newStartIndex,
```

Comment:
why in the `RectangleROIStartEndThreshold`, the startSlice is `startIndex` which is imageIdIndex, but here is different?

Reply:
I believe I grasp the reason why, but just to confirm, this tool differs from the RectangleROITool in that it projects both before and after, whereas the rectangle only projects to after. Is this the intended functionality, or is it purely coincidental?

Reply:
Yes it is intentional, when the physician is going to draw a 3D ROI he is likely to use the center slice to draw the ROI which should propagate upper and lower of the drawn ROI.

It's hard to imagine that the physician is going to to draw a ROI on a upperslice (in which the tumor is not visible if you don't want to miss slices) and expect the ROI to propagate only in lowerslice